pkg/controller/statusmanager/pod_status.go

```diff
@@ -175,54 +215,72 @@ func (status *StatusManager) updateNodeStatus(nodeName string, ready bool, reaso
-	if err != nil {
-		log.Error(err, fmt.Sprintf("Error getting %s for node condition", operatortypes.NsxNodeAgentContainerName))
-	} else if len(pods.Items) == 0 {
-		log.Info("nsx-node-agent not found for node condition")
-		nodes := corev1.NodeList{}
-		err = status.client.List(context.TODO(), &nodes)
+func (status *StatusManager) setNodeConditionFromPod(podName types.NamespacedName, sharedInfo *sharedinfo.SharedInfo, pod *corev1.Pod) {
+	now := time.Now()
+	if pod == nil {
+		pod = &corev1.Pod{}
+		err := status.client.Get(context.TODO(), podName, pod)
 		if err != nil {
-			log.Error(err, "Failed to get nodes for condition updating")
+			log.Error(err, fmt.Sprintf("Error getting %s for node condition", operatortypes.NsxNodeAgentContainerName))
 			return
 		}
-		for _, node := range nodes.Items {
-			status.setNodeNetworkUnavailable(node.ObjectMeta.Name, false, "NSXNodeAgent", "Waiting for nsx-node-agent to be created", sharedInfo)
-		}
-		return
 	}
-	pod := corev1.Pod{}
-	for _, pod = range pods.Items {
-		containerStatus := corev1.ContainerStatus{}
-		ready := true
-		messages := ""
-		nodeName := pod.Spec.NodeName
-		for _, containerStatus = range pod.Status.ContainerStatuses {
-			if containerStatus.State.Running == nil {
-				ready = false
-				if containerStatus.State.Waiting != nil {
-					if containerStatus.State.Waiting.Message == "" {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Waiting.Reason)
-					} else {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Waiting.Message)
-					}
-				} else if containerStatus.State.Terminated != nil {
-					if containerStatus.State.Terminated.Message == "" {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Terminated.Reason)
-					} else {
-						messages = messages + fmt.Sprintf("nsx-node-agent/%s not running: %s. ", containerStatus.Name, containerStatus.State.Terminated.Message)
-					}
+	containerStatus := corev1.ContainerStatus{}
+	ready := true
+	// messages is used to store the error message from container status,
+	// tmpMsgs is for the status that pod just restarted in a short time
+	var messages, tmpMsgs string
+	nodeName := pod.Spec.NodeName
+	var startedAt *time.Time
+	var lastStartedAt *time.Time
+	// when pod is pending, its ContainerStatuses is empty
+	if len(pod.Status.ContainerStatuses) == 0 {
+		ready = false
+		messages = fmt.Sprintf("nsx-node-agent not running: %s.", pod.Status.Phase)
+	}
+	for _, containerStatus = range pod.Status.ContainerStatuses {
+		if containerStatus.State.Running == nil {
+			ready = false
+			if containerStatus.State.Waiting != nil {
+				if containerStatus.State.Waiting.Message == "" {
+					messages = messages + fmt.Sprintf("%s/%s not running: %s. ", pod.Name, containerStatus.Name, containerStatus.State.Waiting.Reason)
```

Comment:
Will `Waiting.Reason` be empty?

Reply:
In my test I didn't see the Waiting.Reason is empty when Waiting is not empty. But I checked https://godoc.org/k8s.io/api/core/v1#ContainerStateWaiting , both Reason and Message are "optional".. OK let's add an "unknown" message if the reason also is empty.